### PR TITLE
Gmg surface bc

### PR DESCRIPTION
**Pull request by tjhei**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -801,14 +802,76 @@ namespace aspect
...
+          continue;
+
+        velocity.reinit(face);
+        velocity.gather_evaluate (src.block(0), EvaluationFlags::values);
+
+        for (unsigned int q = 0; q < velocity.n_q_points; ++q)
+          {
+            const Tensor<1, dim, VectorizedArray<number>> phi_u_i = velocity.get_value(q);
+            const auto &normal_vector = velocity.get_normal_vector(q);
+            const auto stabilization_tensor = cell_data->free_surface_stabilization_term_table(face - face_range.first, q);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:843]: Are you sure face_range.first is correct? Can you check if this is the number of interior faces? This is certainly wrong when multithreading...

File: source/simulator/stokes_matrix_free.cc [hunk 2]
```diff
@@ -1697,6 +1764,114 @@ namespace aspect
...
+        {
+          const double free_surface_theta =
+            sim.mesh_deformation->template get_matching_mesh_deformation_object<MeshDeformation::FreeSurface<dim>>()
+          .get_free_surface_theta();
+
+          if (sim.parameters.include_melt_transport)
+            {
+              // GMG doesn't support melt transport
+              // It won't get here because of the assert in the constructor
+              // AssertThrow(!sim.parameters.include_melt_transport, ExcNotImplemented());
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1784]: Leave this line in.

File: source/simulator/stokes_matrix_free.cc [hunk 3]
```diff
@@ -1697,6 +1764,114 @@ namespace aspect
...
+                          == active_cell_data.free_surface_boundary_indicators.end())
+                        continue;
+
+                      fe_face_values.reinit(simulator_cell, cell_face_pair.second);
+
+                      face_material_inputs.reinit(fe_face_values, simulator_cell, sim.introspection, sim.solution);
+
+                      sim.compute_material_model_input_values(sim.solution, fe_face_values,
+                                                              simulator_cell, false, face_material_inputs);
+                      sim.material_model->evaluate(face_material_inputs, face_material_outputs);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1845]: Do we do averaging of the surface terms in matrix-based?

File: source/mesh_deformation/free_surface.cc [hunk 4]
```diff
@@ -84,7 +84,8 @@ namespace aspect
                 = cell->face(face_no)->boundary_id();
 
               if (tmp_free_surface_boundary_indicators.find(boundary_indicator)
-                  == tmp_free_surface_boundary_indicators.end())
+                  == tmp_free_surface_boundary_indicators.end()
```
[the remark this answers is not in the document]
Comment (zjiaqi2018) [reply source/mesh_deformation/free_surface.cc:87]: I thought the same, but it is executed without it. This assembler is set in the same place as the Stokes assembler.

File: source/simulator/stokes_matrix_free.cc (hunk 1 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:843]: You are right, it's not the number of interior faces.

File: include/aspect/mesh_deformation/interface.h [hunk 5]
```diff
@@ -538,7 +538,7 @@ namespace aspect
     bool
     MeshDeformationHandler<dim>::has_matching_mesh_deformation_object () const
     {
-      for (typename std::map<types::boundary_id, std::vector<std::unique_ptr<Interface<dim>>>>::iterator boundary_id
+      for (typename std::map<types::boundary_id, std::vector<std::unique_ptr<Interface<dim>>>>::const_iterator boundary_id
```
[the remark this answers is not in the document]
Comment (zjiaqi2018) [reply include/aspect/mesh_deformation/interface.h:541]: Will do it soon.

File: source/simulator/stokes_matrix_free.cc (hunk 3 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:1845]: No we don't, I will use harmonic average in the matrix-based one.

**Comment by zjiaqi2018** [comment]
The results seem to be correct now, but GMG requires more iterations. 

image

image

**Comment by tjhei** [comment]
Is it similar if both use cheap iterations only?
This might be because the stabilization is not added to the levels yet. You can confirm by checking theta=0.

**Comment by zjiaqi2018** [comment]
`set Number of cheap Stokes solver steps = 200` for both solvers, the matrix-free solver takes about twice the number of iterations as the matrix-based one:
image

>This might be because the stabilization is not added to the levels yet. You can confirm by checking theta=0.

Yes, we may need to add it to the levels.
When `set Free surface stabilization theta = 0.`:


image

**Comment by tjhei** [comment]
well, this is great progress then!

File: source/mesh_deformation/free_surface.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjhei) [reply source/mesh_deformation/free_surface.cc:87]: But there is no reason to ever add this assembler or execute assembly in the first place. We should leave this as is and check why matrix assembly is being done.

File: source/simulator/stokes_matrix_free.cc [hunk 6]
```diff
@@ -1697,6 +1765,114 @@ namespace aspect
             level_cell_data[level].enable_newton_derivatives = false;
         }
     }
+
+    {
+      // Create active mesh tables to store the product of the pressure perturbation and
+      // the normalized gravity use in the free surface stabilization.
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1771]:
```suggestion
      // the normalized gravity used in the free surface stabilization.
```

File: source/simulator/stokes_matrix_free.cc [hunk 7]
```diff
@@ -1697,6 +1765,114 @@ namespace aspect
...
+        {
+          const double free_surface_theta =
+            sim.mesh_deformation->template get_matching_mesh_deformation_object<MeshDeformation::FreeSurface<dim>>()
+          .get_free_surface_theta();
+
+          if (sim.parameters.include_melt_transport)
+            {
+              // GMG doesn't support melt transport
+              // It won't get here because of the assert in the constructor
+              AssertThrow(!sim.parameters.include_melt_transport, ExcNotImplemented());
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1785]:
```suggestion
              // GMG doesn't support melt transport
              AssertThrow(!sim.parameters.include_melt_transport, ExcNotImplemented());
```
and just remove the whole ``if ... else``.

File: source/simulator/stokes_matrix_free.cc [hunk 8]
```diff
@@ -1697,6 +1765,114 @@ namespace aspect
...
+
+                      face_material_inputs.reinit(fe_face_values, simulator_cell, sim.introspection, sim.solution);
+
+                      sim.compute_material_model_input_values(sim.solution, fe_face_values,
+                                                              simulator_cell, false, face_material_inputs);
+                      sim.material_model->evaluate(face_material_inputs, face_material_outputs);
+
+                      for (unsigned int q = 0; q < n_face_q_points; ++q)
+                        {
+
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1850]:
```suggestion
```

File: source/simulator/stokes_matrix_free.cc [hunk 9]
```diff
@@ -1697,6 +1765,114 @@ namespace aspect
...
+                          const Tensor<1,dim>
+                          gravity = sim.gravity_model->gravity_vector(fe_face_values.quadrature_point(q));
+                          const double g_norm = gravity.norm();
+
+                          // construct the relevant vectors
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1855]:
```suggestion
```

File: source/simulator/stokes_matrix_free.cc [hunk 10]
```diff
@@ -1779,6 +1955,39 @@ namespace aspect
         pressure.integrate_scatter (EvaluationFlags::values,
                                     rhs_correction.block(1));
       }
+
+    if (active_cell_data.apply_stabilization_free_surface_faces)
+      {
+        const unsigned int n_faces_boundary = stokes_matrix.get_matrix_free()->n_boundary_face_batches();
+        const unsigned int n_faces_interior = stokes_matrix.get_matrix_free()->n_inner_face_batches();
+        FEFaceEvaluation<dim,velocity_degree,velocity_degree+1,dim,double>
+        velocity_boundary(*stokes_matrix.get_matrix_free(), true, 0);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1964]: please document ``true`` and ``0``

File: source/simulator/stokes_matrix_free.cc [hunk 11]
```diff
@@ -2421,6 +2630,14 @@ namespace aspect
       additional_data.mapping_update_flags = (update_values | update_gradients |
                                               update_JxW_values | update_quadrature_points);
 
+      if (!sim.mesh_deformation->get_free_surface_boundary_indicators().empty())
+        additional_data.mapping_update_flags_boundary_faces =
+          (update_values  |
+           update_gradients |
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:2636]: do we need the gradients?

**Comment by tjhei** [comment]
Please rebase to master. I just merged the other PR.
Also, can you investigate why ``nonlinear_channel_flow_tractions_Newton_Stokes_GMG`` , ``no_cheap_gmg``, etc fail?
see https://github.com/tjhei/aspect/pull/11/checks?check_run_id=3707043141#step:6:3349

File: source/simulator/stokes_matrix_free.cc (hunk 11 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:2636]: it's copied from matrix-based version, not sure if it's needed in the material model. 
https://github.com/tjhei/aspect/blob/30a5e9658c76e4132e114b5d467f7fc25930455f/source/mesh_deformation/free_surface.cc#L202

Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:2636]: I don't see where you use derivatives, so try it out without.

File: source/mesh_deformation/free_surface.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (zjiaqi2018) [reply source/mesh_deformation/free_surface.cc:87]: I tried putting `if(!(this->is_stokes_matrix_free()))` in front of
https://github.com/tjhei/aspect/blob/30a5e9658c76e4132e114b5d467f7fc25930455f/source/mesh_deformation/free_surface.cc#L177
but it didn't work, it still set the assembler.

File: source/simulator/stokes_matrix_free.cc (hunk 11 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:2636]: you are right!

File: source/mesh_deformation/free_surface.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjhei) [reply source/mesh_deformation/free_surface.cc:87]: Well, that means line 194 is somehow executed. Find out why!
Secondly, we shouldn't even execute the matrix assembly. Find out why that happens.

If in doubt, start a debugger and put a breakpoint there.

File: source/simulator/stokes_matrix_free.cc [hunk 12]
```diff
@@ -1960,8 +1960,11 @@ namespace aspect
       {
         const unsigned int n_faces_boundary = stokes_matrix.get_matrix_free()->n_boundary_face_batches();
         const unsigned int n_faces_interior = stokes_matrix.get_matrix_free()->n_inner_face_batches();
+        //
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1963]: ?

File: source/simulator/stokes_matrix_free.cc [hunk 13]
```diff
@@ -801,14 +802,79 @@ namespace aspect
...
+  void
+  MatrixFreeStokesOperators::StokesOperator<dim, degree_v, number>
+  ::local_apply_boundary_face(const dealii::MatrixFree<dim, number> &data,
+                              dealii::LinearAlgebra::distributed::BlockVector<number> &dst,
+                              const dealii::LinearAlgebra::distributed::BlockVector<number> &src,
+                              const std::pair<unsigned int, unsigned int> &face_range) const
+  {
+    FEFaceEvaluation<dim, degree_v, degree_v + 1, dim, number> velocity(data,
+                                                                        true, /*is_interior_face*/
+                                                                        0 /*number of dof*/);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:829]: This 0 is for selection of the first DoFHandler. Maybe just delete as this is the default?

File: source/simulator/stokes_matrix_free.cc [hunk 14]
```diff
@@ -801,14 +802,78 @@ namespace aspect
...
+  template <int dim, int degree_v, typename number>
+  void
+  MatrixFreeStokesOperators::StokesOperator<dim, degree_v, number>
+  ::local_apply_boundary_face(const dealii::MatrixFree<dim, number> &data,
+                              dealii::LinearAlgebra::distributed::BlockVector<number> &dst,
+                              const dealii::LinearAlgebra::distributed::BlockVector<number> &src,
+                              const std::pair<unsigned int, unsigned int> &face_range) const
+  {
+    FEFaceEvaluation<dim, degree_v, degree_v + 1, dim, number> velocity(data,
+                                                                        true, /*is_interior_face*/);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:828]: Shouldn't this be false? Also remove ,

Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:828]: is interior face of the cell, I should have been more specific.

**Comment by zjiaqi2018** [comment]
GMG failed in the first time step:
```

An error occurred in line <2421> of file </home/jiaqi2/aspect/aspect/source/utilities.cc> in function
    void aspect::Utilities::linear_solver_failed(const string&, const string&, const std::vector<dealii::SolverControl>&, const std::exception&, const MPI_Comm&, const string&)
The violated condition was: 
    false
Additional information: 
    The iterative Stokes solver in
    StokesMatrixFreeHandlerImplementation::solve did not converge.
    
    The initial residual was: nan
    The final residual is: nan
    The required residual for convergence is: nan
    See output_viscoelastic_plate_flexure_gmg/solver_history.txt for the
    full convergence history.
    
    The solver reported the following error:
    
    --------------------------------------------------------
    An error occurred in line <1285> of file
    </hdd1/shared/libs-candi-9.3.1-r2/deal.II-v9.3.1/include/deal.II/lac/solver_gmres.h>
    in function
    void dealii::SolverFGMRES<VectorType>::solve(const MatrixType&,
    VectorType&, const VectorType&, const PreconditionerType&) [with
    MatrixType = aspect::MatrixFreeStokesOperators::StokesOperator<2, 2,
    double>; PreconditionerType =
    aspect::internal::BlockSchurGMGPreconditioner<aspect::MatrixFreeStokesOperators::StokesOperator<2,
    2, double>, aspect::MatrixFreeStokesOperators::ABlockOperator<2, 2,
    double>, aspect::MatrixFreeStokesOperators::MassMatrixOperator<2, 1,
    double>, dealii::PreconditionMG<2,
    dealii::LinearAlgebra::distributed::Vector<double>,
    dealii::MGTransferMatrixFree<2, double> >, dealii::PreconditionMG<2,
    dealii::LinearAlgebra::distributed::Vector<double>,
    dealii::MGTransferMatrixFree<2, double> > >; VectorType =
    dealii::LinearAlgebra::distributed::BlockVector<double>]
    The violated condition was:
    false
    Additional information:
    Iterative method reported convergence failure in step 1. The residual
    in the last step was nan.
    
    This error message can indicate that you have simply not allowed a
    sufficiently large number of iterations for your iterative solver to
    converge. This often happens when you increase the size of your
    problem. In such cases, the last residual will likely still be very
    small, and you can make the error go away by increasing the allowed
    number of iterations when setting up the SolverControl object that
    determines the maximal number of iterations you allow.
    
    The other situation where this error may occur is when your matrix is
    not invertible (e.g., your matrix has a null-space), or if you try to
    apply the wrong solver to a matrix (e.g., using CG for a matrix that
    is not symmetric or not positive definite). In these cases, the
    residual in the last iteration is likely going to be large.
    --------------------------------------------------------
    
--------------------------------------------------------

Aborting!
```

**Comment by zjiaqi2018** [comment]
matrix-based failed in step 40: 
   ```
 The solver reported the following error:
    
    --------------------------------------------------------
    An error occurred in line <1285> of file
    </hdd1/shared/libs-candi-9.3.1-r2/deal.II-v9.3.1/include/deal.II/lac/solver_gmres.h>
    in function
    void dealii::SolverFGMRES<VectorType>::solve(const MatrixType&,
    VectorType&, const VectorType&, const PreconditionerType&) [with
    MatrixType = aspect::internal::StokesBlock; PreconditionerType =
    aspect::internal::BlockSchurPreconditioner<dealii::TrilinosWrappers::PreconditionAMG,
    dealii::TrilinosWrappers::PreconditionBase>; VectorType =
    dealii::TrilinosWrappers::MPI::BlockVector]
    The violated condition was:
    false
    Additional information:
    Iterative method reported convergence failure in step 1000. The
    residual in the last step was 2.62833e+36.
    
    This error message can indicate that you have simply not allowed a
    sufficiently large number of iterations for your iterative solver to
    converge. This often happens when you increase the size of your
    problem. In such cases, the last residual will likely still be very
    small, and you can make the error go away by increasing the allowed
    number of iterations when setting up the SolverControl object that
    determines the maximal number of iterations you allow.
    
    The other situation where this error may occur is when your matrix is
    not invertible (e.g., your matrix has a null-space), or if you try to
    apply the wrong solver to a matrix (e.g., using CG for a matrix that
    is not symmetric or not positive definite). In these cases, the
    residual in the last iteration is likely going to be large.
    --------------------------------------------------------
    
--------------------------------------------------------

Aborting!
```

File: source/simulator/stokes_matrix_free.cc [hunk 15]
```diff
@@ -1779,6 +1954,41 @@ namespace aspect
...
       }
+
+    if (active_cell_data.apply_stabilization_free_surface_faces)
+      {
+        const unsigned int n_faces_boundary = stokes_matrix.get_matrix_free()->n_boundary_face_batches();
+        const unsigned int n_faces_interior = stokes_matrix.get_matrix_free()->n_inner_face_batches();
+
+        FEFaceEvaluation<dim,velocity_degree,velocity_degree+1,dim,double>
+        velocity_boundary(*stokes_matrix.get_matrix_free(),
+                          true /*is interior face of the cell*/);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1965]: this is the default, so you can remove the second arg.

File: source/simulator/assemblers/stokes.cc [hunk 16]
```diff
@@ -335,7 +335,7 @@ namespace aspect
                 {
                   scratch.phi_u[i_stokes] = scratch.finite_element_values[introspection.extractors.velocities].value (i,q);
                   scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection.extractors.pressure].value (i, q);
-                  if (scratch.rebuild_stokes_matrix)
+                  if (scratch.rebuild_stokes_matrix || this->get_parameters().enable_elasticity)
```
Comment (tjhei) [inline source/simulator/assemblers/stokes.cc:338]: Can you please open a PR for this? I think we only need the grads though, not the div.

Comment (zjiaqi2018) [reply source/simulator/assemblers/stokes.cc:338]: add another if?
```
else if (this->get_parameters().enable_elasticity)
   {
       scratch.grads_phi_u[i_stokes] =  scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(i,q);
    }

```

File: benchmarks/test_gmg_free_surface/S20RTS.prm [hunk 17]
```diff
@@ -0,0 +1,32 @@
+include $ASPECT_SOURCE_DIR/cookbooks/S20RTS.prm
```
Comment (tjhei) [inline benchmarks/test_gmg_free_surface/S20RTS.prm:1]: you will need to remove this folder again, I think.

File: include/aspect/mesh_deformation/free_surface.h [hunk 18]
```diff
@@ -108,6 +108,11 @@ namespace aspect
          */
         void parse_parameters (ParameterHandler &prm) override;
 
+        /**
+         * Return the stabilization parameter for the free surface.
+         */
+        double get_free_surface_theta () const;
```
Comment (tjhei) [inline include/aspect/mesh_deformation/free_surface.h:114]: please make a separate PR

File: source/simulator/stokes_matrix_free.cc [hunk 19]
```diff
@@ -1282,12 +1356,6 @@ namespace aspect
     parse_parameters(prm);
     CitationInfo::add("mf");
 
-    if (sim.parameters.mesh_deformation_enabled)
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1285]: can you add an assert that disables mesh_deformation + elasticity for now?

File: source/simulator/stokes_matrix_free.cc [hunk 20]
```diff
@@ -1697,6 +1765,112 @@ namespace aspect
...
+
+              active_cell_data.free_surface_stabilization_term_table.reinit(n_faces_boundary, n_face_q_points);
+
+              for (unsigned int face=n_faces_interior; face<n_faces_boundary + n_faces_interior; ++face)
+                {
+                  const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_active_entries_per_face_batch(face);
+
+                  for (unsigned int i=0; i<n_components_filled; ++i)
+                    {
+                      // the face iterator can be accessed: pair.first->face(pair.second);
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1821]: I don't understand this comment, please delete or fix.

Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:1821]: maybe say what the return value is "(cell_iterator, face_index)"

File: benchmarks/test_gmg_free_surface/S20RTS.prm (hunk 17 above)
Comment (zjiaqi2018) [reply benchmarks/test_gmg_free_surface/S20RTS.prm:1]: remove this prm? How about S20RTS_gmg.prm?

File: source/simulator/stokes_matrix_free.cc (hunk 19 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:1285]: Like this?
`AssertThrow(!(sim.parameters.mesh_deformation_enabled && sim.parameters.enable_elasticity), ExcNotImplemented());`
then I will need to remove  free_surface_viscous_cylinder_2D_loading_gmg.prm

Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:1285]: See what @naliboff had to say on slack.

File: benchmarks/test_gmg_free_surface/S20RTS.prm (hunk 17 above)
Comment (tjhei) [reply benchmarks/test_gmg_free_surface/S20RTS.prm:1]: You added the folder ``benchmarks/test_gmg_free_surface``. I think it doesn't  make sense to keep it around. It is not really a "benchmark". We could add some files next to existing ones (bla_gmg.prm next to bla.prm).

Comment (zjiaqi2018) [reply benchmarks/test_gmg_free_surface/S20RTS.prm:1]: I see. Do you want to keep the matrix-based version to make it easier to compare? (bla_amp.prm and bla_gmg.prm)
We changed some parameters in the matrix-based prm to get a better comparison with the matrix-free one.
